Add a Service Operator for PostgreSQL databases

issyl0 opened the pull request:
- Our users (in this case DCS) want to be able to provision Postgres databases. We want to create a Postgres custom resource that will provide a Postgres database for them (ie Amazon RDS).
- This is a no-op operator which allows our users to specify a config file of the form:
```yaml
kind: Postgres
spec:
  aws:
    instanceType: db.m5.large
  local:
    image: postgres:latest
```
- This will eventually (in the next few stories) create an RDS instance of type `db.m5.large` in AWS, when applied.

https://trello.com/c/Ob1NeQeV/61-create-and-deploy-basic-no-op-postgres-operator

Co-authored-by: Rafal Proszowski <rafal.proszowski@digital.cabinet-office.gov.uk>

issyl0 committed:
kubebuilder init --domain gsp.k8s.io/v1beta1 && make all

- Run through the first few steps of https://book.kubebuilder.io/quick-start.html.

issyl0 committed:
kubebuilder create api --group database --version v1beta1 --kind Postgres

- This generates a load of boilerplate.
- We thought about calling this "service", but that's overloaded.
- We can have "database" and "queue" groups instead.

issyl0 committed:
Allow specifying InstanceType and retrieving Status in the CRD

- This will look something like:

  ```yaml
  kind: Postgres
  spec:
    aws:
      instanceType: db.m5.large
    local:
      image: postgres:latest
  ```

Co-authored-by: Rafal Proszowski <rafal.proszowski@digital.cabinet-office.gov.uk>

issyl0 committed:
Run `make` to generate the CRD for our InstanceType etc. changes

Co-authored-by: Rafal Proszowski <rafal.proszowski@digital.cabinet-office.gov.uk>

issyl0 committed:
Log what options the Postgres operator expects

Co-authored-by: Rafal Proszowski <rafal.proszowski@digital.cabinet-office.gov.uk>

issyl0 commented on components/service-operator/Dockerfile:
```diff
@@ -0,0 +1,25 @@
+COPY controllers/ controllers/
+
+# Build
+RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 GO111MODULE=on go build -a -o manager main.go
```
I suppose `GOOS=linux` shouldn't be changed to `GOOS=darwin` because we'll be running this on GSP on Linux. :apple: :penguin:

philandstuff commented:
> `    instanceType: m5d.large`

I don't see `m5d.large` in the RDS instance type list.  I see `db.m5.large` though?

More generally, does this deserve a README of its own?

issyl0 commented:
@paroxp I think your rebase went wrong. :-)

issyl0 committed:
Add a basic README saying what the plan is

issyl0 commented:
I added a _very basic_ README, we can iterate it later. And I changed `m5d.large` to `db.m5.large` in all the places.

paroxp commented:
> I don't see `m5d.large` in the RDS instance type list. I see `db.m5.large` though?

You are right, we're just not at the position to be applying any configuration, meaning we are likely to play with this a little more.

> More generally, does this deserve a README of its own?

Agree, README will come when we grow the thing. At the moment, the only notable "paragraph" would be how to add a new database service.

paroxp reviewed (changes requested):
Let's include Pipeline changes as part of this PR :)

paroxp committed:
Add service-operator release job

We are developing a generic service-operator for GDS, and we'd like to
follow our current approach of CI/CD the solution. Extending the
existing pipeline, should assure we're on the same path.

Co-authored-by: Issy Long <isabell.long@digital.cabinet-office.gov.uk>
Co-authored-by: Rafal Proszowski <rafal.proszowski@digital.cabinet-office.gov.uk>

paroxp committed:
Add tests to the build

We'd like to be running tests before we build the operator. We're adding
the test task into the job, so that the build is only executed after the
tests pass.

Co-authored-by: Issy Long <isabell.long@digital.cabinet-office.gov.uk>
Co-authored-by: Rafal Proszowski <rafal.proszowski@digital.cabinet-office.gov.uk>
